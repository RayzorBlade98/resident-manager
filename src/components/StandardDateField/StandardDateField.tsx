import { DateField } from '@mui/x-date-pickers';
import { Dayjs } from 'dayjs';
import React from 'react';

interface StandardDateFieldProps {
  /**
   * Label of the datepicker field
   */
  label: string;

  /**
   * Callback when selecting a date
   * @param date selected date
   */
  onChange: (date: Date) => void;

  /**
   * Whether the date field should be shown in error state
   */
  error?: boolean;
}

/**
 * Generic date field component
 */
function StandardDateField(props: StandardDateFieldProps): JSX.Element {
  return (
    <DateField
      required
      label={props.label}
      format="DD.MM.YYYY"
      onChange={(date) => props.onChange((date as Dayjs).toDate())}
      color={props.error ? 'error' : 'primary'}
    />
  );
}

export default StandardDateField;
