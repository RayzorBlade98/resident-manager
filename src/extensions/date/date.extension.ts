declare interface Date {
  /**
   * Converts the date to a new date in UTC timezone
   */
  toUTC(): Date;

  /**
   * Converts the date to 'DD.MM.YYYY' format string
   */
  toPreferredString(): string;
}

function toUTC(this: Date): Date {
  return new Date(
    Date.UTC(this.getFullYear(), this.getMonth(), this.getDate(), 0, 0, 0, 0),
  );
}
Date.prototype.toUTC = toUTC;

function toPreferredString(this: Date): string {
  const day: number = this.getDate();
  const month: number = this.getMonth() + 1;
  const year = String(this.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
Date.prototype.toPreferredString = toPreferredString;
