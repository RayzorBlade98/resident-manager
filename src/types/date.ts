/**
 * Enum containing all months as `string`
 */
export enum Month {
  January = 'Januar',
  Febuary = 'Februar',
  March = 'MÃ¤rz',
  April = 'April',
  May = 'Mai',
  June = 'Juni',
  July = 'Juli',
  August = 'August',
  September = 'September',
  October = 'Oktober',
  November = 'November',
  December = 'Dezember',
}

/**
 * Type representing a year
 */
export type Year = number;

/**
 * Information about a specific month and year
 */
export interface MonthYear {
  month: Month;
  year: Year;
}

/**
 * Type of the string representation of a `MonthYear`
 *
 * Format: "[month] [year]"
 */
export type MonthYearString = `${Month} ${Year}`;

/**
 * Type of the string representation of a date including day, month and year
 *
 * Format: `dd.mm.yyyy`
 */
export type DateString = `${string}.${string}.${string}`;

/**
 * Type of the date string representation returned by the date input field
 *
 * Format: `yyyy-mm-dd`
 */
export type DateInputString = `${string}-${string}-${string}`;

/**
 * Class that provides utility functions for `MonthYear` objects
 */
export abstract class MonthYearUtils {
  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param monthYear `MonthYear` object to which the months should be added
   * @param addedMonths Number of months that should be added (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static addMonths(monthYear: MonthYear, addedMonths = 1): MonthYear {
    if (addedMonths < 0) {
      // Add negative amount of months == subtract months
      return MonthYearUtils.subtractMonths(monthYear, addedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.December) {
      // Year change
      newMonthYear.month = Month.January;
      newMonthYear.year += 1;
    } else {
      // Add one month to the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex + 1];
    }

    // Call `addMonths` recursive until all months got addded
    if (addedMonths !== 1) {
      return MonthYearUtils.addMonths(newMonthYear, addedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Checks if two `MonthYear` objects are equal
   * @param a first `MonthYear` object to compare
   * @param b second `MonthYear` object to compare
   * @returns `true` if both the month and year are the same, else `false`
   */
  public static areEqual(a: MonthYear, b: MonthYear): boolean {
    return a.month === b.month && a.year === b.year;
  }

  /**
   * Compares two `MonthYear` objects
   * @param a First `MonthYear` object to compare
   * @param b Second `MonthYear` object to compare
   * @returns
   *
   * - 0 if both object have the same month and year
   * - <0 if the first object contains a date that's earlier than the second object
   * - \>0 if the first object contains a date that's later than the second object
   */
  public static compare(a: MonthYear, b: MonthYear): number {
    if (a.year !== b.year) return a.year - b.year;

    const months = Object.values(Month);
    const aMonthIndex = months.findIndex((m: Month) => m === a.month);
    const bMonthIndex = months.findIndex((m: Month) => m === b.month);
    return aMonthIndex - bMonthIndex;
  }

  /**
   * Returns a `MonthYear` object that contains information of the current month and year
   * @returns `MonthYear` object for the current date
   */
  public static getCurrentMonthYear(): MonthYear {
    const date: Date = new Date();
    const monthIndex: number = date.getUTCMonth();
    const month: Month = Object.values(Month)[monthIndex] as Month;
    const year: Year = date.getUTCFullYear();

    return {
      month,
      year,
    };
  }

  /**
   * Parses a string representation of a `MonthYear` object to its respective object
   * @param monthYear string representation of a `MonthYear` object that should be parsed
   * @returns `MonthYear` object that contains the information of the specified `string`
   */
  public static parseString(monthYear: MonthYearString): MonthYear {
    const splits = monthYear
      .split(/(\s+)/)
      .filter((s: string) => s.trim().length > 0);
    return {
      month: splits[0] as Month,
      year: Number(splits[1]),
    };
  }

  /**
   * Subtracts the specified number of months from the given `MonthYear` object
   * @param monthYear `MonthYear` object from which the months should be subtraced
   * @param subtractedMonths Number of months that should be subtracted (can be negative)
   * @returns new `MonthYear` object with the adjusted month (and year)
   */
  public static subtractMonths(
    monthYear: MonthYear,
    subtractedMonths = 1,
  ): MonthYear {
    if (subtractedMonths < 0) {
      // Subtract negative amount of months == add months
      return MonthYearUtils.addMonths(monthYear, subtractedMonths * -1);
    }

    const newMonthYear = { ...monthYear };
    if (newMonthYear.month === Month.January) {
      // Year change
      newMonthYear.month = Month.December;
      newMonthYear.year -= 1;
    } else {
      // Subtract one month from the object
      const months = Object.values(Month);
      const oldMonthIndex = months.findIndex(
        (m: Month) => m === newMonthYear.month,
      );
      newMonthYear.month = months[oldMonthIndex - 1];
    }

    // Call `subtractMonths` recursive until all months got subtracted
    if (subtractedMonths !== 1) {
      return MonthYearUtils.subtractMonths(newMonthYear, subtractedMonths - 1);
    }
    return newMonthYear;
  }

  /**
   * Converts a `MonthYear` to its string representation
   * @param monthYear `MonthYear` that should be converted
   * @returns string representation of the specified `MonthYear` object with the format "[month] [year]"
   */
  public static toString(monthYear: MonthYear): MonthYearString {
    return `${monthYear.month} ${monthYear.year}`;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    const timespan = [{ ...start }];

    // Whether to add a month each loop or subtract one
    const direction = Math.sign(MonthYearUtils.compare(end, start));

    if (MonthYearUtils.areEqual(start, end)) return timespan;

    for (
      let next = MonthYearUtils.addMonths(start, direction);
      !MonthYearUtils.areEqual(next, end);
      next = MonthYearUtils.addMonths(next, direction)
    ) {
      timespan.push(next);
    }
    timespan.push({ ...end });
    return timespan;
  }
}

/**
 * Class that provides utility functions for dates containing day, month and year.
 */
export abstract class DateUtils {
  /**
   * Converts a date string from `yyyy-mm-dd` to `dd.mm.yyyy` format
   * @param date Date in `yyyy-mm-dd` format
   * @returns Converted date in `dd.mm.yyyy` format
   */
  public static convertDateInputString(date: DateInputString): DateString {
    const splits = date.split('-');
    return `${splits[2]}.${splits[1]}.${splits[0]}`;
  }

  /**
   * Converts a date string from `dd.mm.yyyy` to `yyyy-mm-dd` format
   * @param date Date in `dd.mm.yyyy` format
   * @returns Converted date in `yyyy-mm-dd` format
   */
  public static convertToDateInputString(date: DateString): DateInputString {
    const splits = date.split('.');
    return `${splits[2]}-${splits[1]}-${splits[0]}`;
  }

  /**
   * Returns a date string containing the current date
   * @returns Date string containing the current date in `dd.mm.yyyy` format
   */
  public static getCurrentDate(): DateString {
    const date: Date = new Date();
    const day: number = date.getUTCDate();
    const month: number = date.getUTCMonth() + 1;
    const year = String(date.getUTCFullYear());

    const dayPrefix: string = day < 10 ? '0' : '';
    const monthPrefix: string = month < 10 ? '0' : '';

    return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
  }
}
