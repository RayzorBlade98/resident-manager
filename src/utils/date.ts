// Remove after adding another export:
/* eslint-disable import/prefer-default-export */

/**
 * Converts the given date to its string representaion (format: DD.MM.YYYY)
 * @param date date that should be converted
 * @returns string representation of the date
 */
export function dateToString(date: Date): string {
  const day: number = date.getDate();
  const month: number = date.getMonth() + 1;
  const year = String(date.getFullYear());

  const dayPrefix: string = day < 10 ? '0' : '';
  const monthPrefix: string = month < 10 ? '0' : '';

  return `${dayPrefix}${String(day)}.${monthPrefix}${String(month)}.${year}`;
}
