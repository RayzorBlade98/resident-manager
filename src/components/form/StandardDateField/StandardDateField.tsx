import { DateField } from '@mui/x-date-pickers';
import dayjs, { Dayjs } from 'dayjs';
import React from 'react';
import '_/extensions/date/date.extension';

interface StandardDateFieldProps {
  /**
   * Label of the datepicker field
   */
  label: string;

  /**
   * Current date value of the field
   */
  value: Date | undefined;

  /**
   * Callback when selecting a date
   * @param date selected date
   */
  onChange: (date: Date | undefined) => void;

  /**
   * Error message displayed below the field
   */
  errorMessage?: string;

  /**
   * Id of the date field
   */
  id?: string;

  /**
   * Whether the field is required or not
   */
  required?: boolean;
}

/**
 * Generic date field component
 */
function StandardDateField(props: StandardDateFieldProps): JSX.Element {
  const onChange = (date: Dayjs | null) => {
    let parsedDate: Date | undefined = date?.toDate();
    if (parsedDate) {
      if (Number.isNaN(parsedDate.valueOf())) {
        parsedDate = undefined;
      } else {
        parsedDate = parsedDate.toUTC();
      }
    }
    props.onChange(parsedDate);
  };

  return (
    <DateField
      required={props.required}
      label={props.label}
      format="DD.MM.YYYY"
      value={props.value ? dayjs(props.value) : undefined}
      onChange={onChange}
      slotProps={{
        textField: {
          helperText: props.errorMessage,
          error: !!props.errorMessage,
          id: props.id,
        },
      }}
    />
  );
}

export default StandardDateField;
