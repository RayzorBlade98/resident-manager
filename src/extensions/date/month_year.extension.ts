import './date.extension';

const MONTH_NAMES = [
  'Januar',
  'Februar',
  'MÃ¤rz',
  'April',
  'Mai',
  'Juni',
  'Juli',
  'August',
  'September',
  'Oktober',
  'November',
  'Dezember',
];

/**
 * Date object that represents a month and a year
 */
class MonthYear extends Date {
  /**
   * Creates a new `MonthYear` object
   * @param month month of the new object. If undefined the current month will be used.
   * @param year year of the new object. If undefined the current year will be used.
   */
  constructor(month?: number, year?: number) {
    const now = new Date();
    const utc = new Date(
      year !== undefined ? year : now.getFullYear(),
      month !== undefined ? month : now.getMonth(),
      1,
    ).toUTC();
    super(utc.getTime());
  }

  /**
   * Converts a `Date` object to a `MonthYear`
   * @param date Date that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromDate(date: Date): MonthYear {
    return new MonthYear(date.getMonth(), date.getFullYear());
  }

  /**
   * Converts a date string to a `MonthYear`
   * @param date Date string that should be converted
   * @returns `MonthYear` with the same month and year as `date`
   */
  public static fromString(date: string): MonthYear {
    return MonthYear.fromDate(new Date(date));
  }

  /**
   * Adds the specified number of months to the given `MonthYear` object
   * @param toAdd Number of months that should be added (can be negative)
   * @returns this month year
   */
  public addMonths(toAdd = 1): MonthYear {
    this.setMonth(this.getMonth() + toAdd);
    return this;
  }

  /**
   * Creates a copy this object.
   * @returns the copy with the same month and year
   */
  public clone(): MonthYear {
    return new MonthYear(this.getMonth(), this.getFullYear());
  }

  /**
   * Checks whether this object has the same month and year as anoter `MonthYear`
   * @param other true if both the month and year are the same, else false
   */
  public equals(other: MonthYear): boolean {
    return this >= other && this <= other;
  }

  /**
   * Creates a list of `MonthYear` objects that contains each month
   * between the two specified `MonthYear` objects (inclusive these months)
   * @param start Start of the timespan
   * @param end End of the timespan
   * @returns List of `MonthYear` objects with all months between `start` and `end`
   */
  public static timespan(start: MonthYear, end: MonthYear): MonthYear[] {
    // Whether to add a month each loop or subtract one
    const direction = end > start ? 1 : -1;

    const timespan: MonthYear[] = [start.clone()];
    let next = start;
    while (!end.equals(next)) {
      next = next.clone();
      next.addMonths(direction);
      timespan.push(next);
    }

    return timespan;
  }

  /**
   * Generates a string representation in the format `<month name> <year>`
   * @returns string representation of this object
   */
  public toString(): string {
    return `${MONTH_NAMES[this.getMonth()]} ${this.getFullYear()}`;
  }
}

export default MonthYear;
