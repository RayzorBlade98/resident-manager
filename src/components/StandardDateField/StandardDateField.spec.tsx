import { RenderResult, render } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import StandardDateField from './StandardDateField';
import RecoilTestWrapper from '_tests/__test_utils__/RecoillTestWrapper';
import { dateToUTC } from '_utils/date';

describe('StandardDateField', () => {
  const onChangeMock = jest.fn();
  const labelText = 'Testlabel';
  let value: Date | undefined;
  let errorMessage: string | undefined;
  let renderResult: RenderResult;

  function renderComponent(): void {
    renderResult = render(
      <RecoilTestWrapper>
        <StandardDateField
          label={labelText}
          value={value}
          onChange={onChangeMock}
          errorMessage={errorMessage}
        />
      </RecoilTestWrapper>,
    );
  }

  beforeEach(() => {
    errorMessage = 'Testerror';
    value = dateToUTC(new Date(2023, 4, 27));
    renderComponent();
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('should display label', () => {
    // Assert
    const label = renderResult.container.querySelector('label');
    expect(label?.textContent).toEqual(labelText);
  });

  test('should display error message', () => {
    // Assert
    const error = renderResult.container.querySelector('p');
    expect(error?.textContent).toEqual(errorMessage);
  });

  test('should display the value', () => {
    // Assert
    const field = renderResult.container.querySelector(
      'input',
    ) as HTMLInputElement;
    expect(field.value).toBe('27.05.2023');
  });

  test('should call onChange for valid input', async () => {
    // Arrange
    const user = userEvent.setup();
    // Act
    const field = renderResult.getByRole('textbox');
    await user.type(field, '25121998');

    // Assert
    expect(onChangeMock).toHaveBeenLastCalledWith(
      dateToUTC(new Date(1998, 11, 25)),
    );
  });

  test('should call onChange for invalid input', async () => {
    // Arrange
    const user = userEvent.setup();
    // Act
    const field = renderResult.getByRole('textbox');
    await user.click(field);
    await user.keyboard('{Backspace}');

    // Assert
    expect(onChangeMock).toHaveBeenLastCalledWith(undefined);
  });

  test('should match snapshot', () => {
    // Arrange
    errorMessage = undefined;
    renderComponent();

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });

  test('should match snapshot (with error)', () => {
    // Arrange
    value = undefined;
    renderComponent();

    // Assert
    expect(renderResult.container).toMatchSnapshot();
  });
});
