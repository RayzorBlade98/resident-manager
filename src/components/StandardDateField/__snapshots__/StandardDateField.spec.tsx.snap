// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StandardDateField should match snapshot (with error) 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-error MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="false"
      for=":r8:"
      id=":r8:-label"
    >
      Testlabel
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary Mui-error MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-describedby=":r8:-helper-text"
        aria-invalid="true"
        autocomplete="off"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r8:"
        inputmode="text"
        placeholder="TT.MM.JJJJ"
        type="text"
        value=""
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-yjsfm1"
        >
          <span>
            Testlabel
          </span>
        </legend>
      </fieldset>
    </div>
    <p
      class="MuiFormHelperText-root Mui-error MuiFormHelperText-sizeMedium MuiFormHelperText-contained css-1wc848c-MuiFormHelperText-root"
      id=":r8:-helper-text"
    >
      Testerror
    </p>
  </div>
</div>
`;

exports[`StandardDateField should match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r6:"
      id=":r6:-label"
    >
      Testlabel
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        autocomplete="off"
        class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r6:"
        inputmode="text"
        placeholder="TT.MM.JJJJ"
        type="text"
        value="27.05.2023"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-14lo706"
        >
          <span>
            Testlabel
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
